ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/transaction/jts/jca/TransactionImporterImple.java

```diff
@@ -159,7 +169,10 @@ public void removeImportedTransaction (Xid xid) throws XAException
 		if (xid == null)
 			throw new IllegalArgumentException();
 
-		_transactions.remove(new SubordinateXidImple(xid));
+		AtomicReference<SubordinateTransaction> remove = _transactions.remove(new SubordinateXidImple(xid));
+		if (remove != null) {
+			TransactionImple.removeTransaction((com.arjuna.ats.internal.jta.transaction.jts.TransactionImple) remove.get());
+		}
```

Comment:
I know I have already reviewed the other two PRs but I am wondering if there is a race here. When we first import a new xid we initially put an AtomicReference in the map with a null value:

https://github.com/jbosstm/narayana/blob/master/ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/transaction/jts/jca/TransactionImporterImple.java#L180

Therefore it can happen that this remove.get() call could be null and if this is the case then the TransactionImple.removeTransaction() call will blow up.